### DISCUSSION TITLE
Is it possible to self-host my own instance of SharpLab? Has anyone done this? Any guides available?

### DISCUSSION THREAD

**Discussion by fsociety3765** [ashmind/SharpLab#952] [Q&A]

**Comment by fsociety3765** [comment]
It's not immediately obvious how I would go about setting this up. 

Are the build artifacts Docker images? If so, could I spin this up with Docker Compose?

I would want to self-host this on an Ubuntu Server.

**Comment by ashmind** [comment]
Thanks for your question! There are no guides that I am aware of, and the application is not containerized.
However if you have an Azure environment, you can use ARM templates at https://github.com/ashmind/sharplab-infrastructure to automatically set up most of the infrastructure.

It should not be very hard to Dockerize most of it, but it's not something I have focused on yet.

Note that current implementation of "Run" mode is not Linux-compatible at the moment (though it's planned eventually).

**Reply by atcarter714** [reply]
What do you suggest as to the easiest way to run it? It's either that, or I'm gonna have to write my own service for a Blazor WASM app I'm going to host on Azure. I really just want the compilation service, for JIT, ASM, etc in each direction (C# to C# is also useful to see what extra stuff is generated by the compiler) and doesn't really matter where I have to host or run it I guess. As long as its reachable and code in returns code out.

I'd also like to make a way to compile only a _fragment_ of C# code, which doesn't worry about assembly references, whether or not identifiers or types are defined or anything beyond the legal syntax and allowed usage (if there are types, references, etc it doesn't understand it just uses a symbol in the IL and adds a comment). But you could give it just a single C# statement like `int n = 0 + 1;` and get the resulting IL back, no other "noise" included. Maybe that should be one possible mode ... "full" mode will generate the whole program, even the .assembly blocks, "class" mode will generate things to the class/type level in IL, "method" mode would generate only a single method in IL and "statement" mode would do as I just described and turn fragments of C# into fragments of the target.

I'm planning to make some tools, and I figure this project might be able to save a good bit of time. I've written my own Roslyn compiler-as-a-service apps before, but a really good one would be quite a bit of work, and if this could save me some time I'd gladly do it ...

**Comment by DoIev** [comment]
Hey, Following @fsociety3765 's question, I've tried to run my own instance of SharpLab on my computer, without success. I keep getting npm errors on `sl setup` stage bacuse `npm ci` on the WebApp failes. any help? node version 18, dotnet version 7
